TM PDB CCD Contribution

jaclark5 opened the pull request:
## New Submission Checklist

- [X] Created a new folder in the submissions directory containing the dataset
- [X] Added `README.md` describing the dataset see [here](https://github.com/openforcefield/qca-dataset-submission/tree/master/submissions/2020-03-26-OpenFF-Gen-2-Torsion-Set-6-supplemental-2) for examples
- [X] All files used to produce the dataset are included with a description
- [NA] Dataset follows the QCSubmit schema defined for [Datasets](https://github.com/openforcefield/qcsubmit/blob/56680a7d3298b5d8962edcb840b0fdb34558c053/qcsubmit/datasets.py#L340), [OptimizationDatasets](https://github.com/openforcefield/qcsubmit/blob/56680a7d3298b5d8962edcb840b0fdb34558c053/qcsubmit/datasets.py#L1062) and [TorsionDriveDatasets](https://github.com/openforcefield/qcsubmit/blob/56680a7d3298b5d8962edcb840b0fdb34558c053/qcsubmit/datasets.py#L1225)
- [NA] Dataset filename matches pattern `dataset*.json`; may feature a compression extension, such as `.bz2`
- [X] A PDF depicting the molecules is attached, in the case of torsiondrives this should include the highlighting of the central bond, this can be done automatically using [qcsubmit](https://github.com/openforcefield/qcsubmit/blob/56680a7d3298b5d8962edcb840b0fdb34558c053/qcsubmit/datasets.py#L854). 
- [NA] QCSubmit validation passed
- [ ] Made a new dataset entry in the mapping table in repository `README.md`
- [ ] Ready to submit!


## Compute Expansion Checklist

- [ ] Compute expansion filename matches pattern `compute*.json`; may feature a compression extension, such as `.bz2`
- [ ] QCSubmit validation passed
- [ ] Ready to submit!

lilyminium reviewed (commented):
Thanks Jen, I had a quick look. I think my comments on sanitize_tm are largely minor since a lot of the code there seems extremely specific for this dataset.

I've got a few questions comments on the notebook too that I'll leave here, in order of appearance via cell:

- Is `is_isomorphic` checking molecules properly? It looks like it checks whether or not atom indices are bonded to each other, so would firstly rely on atom order being the same, secondly doesn't check element, and thirdly doesn't check bond order. 

- the `rdkit_to_smiles` function: do you want to be adding Hs if there are no Hs in the geometry?

```
 if (
        conformer_list 
        and mol_dict["charge"] != qc_input[conformer_list[0]]["charge"]
        and symbols != qc_input[conformer_list[0]]["symbols"]
    ):
    for other_name in conformer_list:
            conformers_dict[other_name].append(name)
```

- I don't quite understand what is happening in the code snippet above; if another geometry is found, but the charge and the elements are different, it's added to the conformers_dict? Surely if these are different it *shouldn't* be a conformer?

- On dataset creation, there's a space in the "short description" key -- that should probably be "short_description".

jaclark5 commented:
> Is `is_isomorphic` checking molecules properly? It looks like it checks whether or not atom indices are bonded to each other, so would firstly rely on atom order being the same, secondly doesn't check element, and thirdly doesn't check bond order.

You're right it should check atom symbol instead of atom index. I've updated that. I don't think it should check bond order for this case because the ligands are conjugated and we've seen that RDKit with waffle about whether it's aromatic and datively coordinated to the metal center to if it's bonded to the metal center with strict single and double bonds in the ligand. I'm mostly concerned about the symbol graph.

> the rdkit_to_smiles function: do you want to be adding Hs if there are no Hs in the geometry?

Good point, it's removed.

> I don't quite understand what is happening in the code snippet above; if another geometry is found, but the charge and the elements are different, it's added to the conformers_dict? Surely if these are different it shouldn't be a conformer?

You're right, I updated to:
```
    for conf_name in conformer_list:
        if mol_dict["charge"] == qc_input[conf_name]["charge"] and set(symbols) != set(qc_input[conf_name]["symbols"]):
            conformers_dict[conf_name].append(name)
            conformers_dict[name].append(conf_name)
```

> On dataset creation, there's a space in the "short description" key -- that should probably be "short_description".

Thanks!

jaclark5 commented on submissions/2025-03-12-PDB-CCD-Chemical-Space-A-Optimization-Dataset-v0.0/sanitize_tm.py:
```diff
@@ -0,0 +1,536 @@
+        mol (rdkit.Chem.rdchem.Mol): Resolved Ligand molecule 
+    """
+    
+    new_code = False
+    ligand_mol = Chem.RWMol(ligand_mol)
+    if not new_code:
+        total_charge_after, hanging_bonds_after, charged_atoms_after = assess_atoms(ligand_mol)
+        if verbose:
+            print(f"Total ligand charge: {total_charge_after}, N_Hanging Bonds {hanging_bonds_after}")
+            print(f"Charged atom info:")
+            for x,y in charged_atoms_after.items():
+                print(f"    {x}: {y}")
+        return total_charge_after, hanging_bonds_after, charged_atoms_after, ligand_mol
```
Good point, I cleaned out all the code that isn't actually used.

jaclark5 commented on submissions/2025-03-12-PDB-CCD-Chemical-Space-A-Optimization-Dataset-v0.0/sanitize_tm.py:
```diff
@@ -0,0 +1,536 @@
+    symbol = metal.GetSymbol()
+    c_atoms = []
+    for b in metal.GetBonds():
+        carbon = b.GetBeginAtom() if b.GetBeginAtom().GetSymbol() != symbol else b.GetEndAtom()
```
It is an assumption but only if the metal has 10 bonds. That really only happens when there is an aromatic ring above and below it. I added something to raise an error if the center isn't iron.

jaclark5 commented on submissions/2025-03-12-PDB-CCD-Chemical-Space-A-Optimization-Dataset-v0.0/sanitize_tm.py:
```diff
@@ -0,0 +1,536 @@
+        c_atoms.append(carbon.GetIdx())
+        for bc in carbon.GetBonds():
+            tmp_atm = bc.GetBeginAtom() if bc.GetBeginAtomIdx() != carbon.GetIdx() else bc.GetEndAtom()
+            if assert_same_ring(mol, carbon.GetIdx(), tmp_atm.GetIdx()):
+                bc.SetBondType(Chem.BondType.AROMATIC)
```
two parallel hexane's wouldn't both be bonded to Fe at the same time. I added an error catch to ensure that before getting to this point it's ensured that the metal center is Fe and has 10 bonds, the presence of which are preventing hydrogens to be added to the ring.

This code snippet is needed far a particular structure where the benzene ring over an Fe atom is then connected to another aromatic ring.

jaclark5 commented on submissions/2025-03-12-PDB-CCD-Chemical-Space-A-Optimization-Dataset-v0.0/sanitize_tm.py:
```diff
@@ -0,0 +1,536 @@
+    """
+    tm_ox = None
+    mol = Chem.DeleteSubstructs(mol, Chem.MolFromSmarts("[#0]"))
+    Chem.AddHs(mol, addCoords=True, explicitOnly=False)
```
I removed this line because of your previous comment about the hydrogens being added when making a smiles string.

jaclark5 commented on submissions/2025-03-12-PDB-CCD-Chemical-Space-A-Optimization-Dataset-v0.0/sanitize_tm.py:
```diff
@@ -0,0 +1,536 @@
+    coordinating_atoms = np.nonzero(Chem.rdmolops.GetAdjacencyMatrix(mol)[tmc_idx, :])[
+        0
+    ]
+
```
I don't think that quite fits but I added more comments.

lilyminium commented:
Thanks Jen!!

Maybe I'm missing something but I don't completely follow the logic of `set(symbols) != set(qc_input[conf_name]["symbols"])` still; wouldn't you want the symbols to match between conformers?

Either way I think my comments were relatively minor  or metadata related (which can always be corrected post submission). I probably won't have time to review deeply before the end of your work day but your code even pre review would have generated a good dataset, so please don't let me block you from submitting!

jaclark5 commented:
> Thanks Jen!!
> 
> Maybe I'm missing something but I don't completely follow the logic of `set(symbols) != set(qc_input[conf_name]["symbols"])` still; wouldn't you want the symbols to match between conformers?
> 
> Either way I think my comments were relatively minor or metadata related (which can always be corrected post submission). I probably won't have time to review deeply before the end of your work day but your code even pre review would have generated a good dataset, so please don't let me block you from submitting!

Yup, doesn't do much, the new graph comparison covers that though. I haven't pushed the changes because I can't get the FractalSnowflake to work.

jaclark5 commented:
> The "geomeTRIC run_json error" has been seen before in a previous dataset, and is due to an upstream issue in the QCEngine wrapper for xTB: [#146 (comment)](https://github.com/openforcefield/qca-dataset-submission/pull/146#issuecomment-730038110)

The following record ids have this error: [145252005, 145252179, 145252011, 145251988, 145251973, 145252061, 145252320, 145252198, 145252188, 145252194, 145252289]

jaclark5 commented:
There are 97 error (weirdly all with a multiplicity of 3) with an error from QCFractal being unable to read the stdout from the xtb subprocess. A [PR in QCFractal](https://github.com/MolSSI/QCFractal/pull/923) will make this error more meaningful. The following record ids suffer from this error: [145252038, 145252145, 145251991, 145251936, 145252089, 145252090, 145252047, 145252094, 145252049, 145252000, 145252097, 145252052, 145252001, 145252056, 145252003, 145252158, 145251956, 145252105, 145252063, 145251962, 145252065, 145252112, 145252170, 145252172, 145252115, 145252071, 145252020, 145252175, 145251977, 145252124, 145252030, 145251981, 145252127, 145252079, 145252081, 145252035, 145252134, 145252084, 145252037, 145251987, 145252141, 145252041, 145251990, 145252088, 145252045, 145251993, 145251939, 145251997, 145251943, 145252093, 145251999, 145251947, 145252150, 145252051, 145251950, 145252152, 145252154, 145251952, 145252101, 145252060, 145252241, 145252326, 145252285, 145252202, 145252204, 145252330, 145252249, 145252207, 145252252, 145252210, 145252333, 145252213, 145252335, 145252296, 145252255, 145252337, 145252298, 145252301, 145252218, 145252304, 145252267, 145252307, 145252185, 145252309, 145252187, 145252273, 145252189, 145252315, 145252276, 145252317, 145252279, 145252234, 145252193, 145252281, 145252237, 145252196, 145252321]
